Add an "opt-out" global switch to retain second fetch

rishabhBudhouliya opened the pull request:
## [PR-904](https://github.com/jenkinsci/git-plugin/pull/904) Fix redundant fetch

With the latest fix on redundant fetch issue [JENKINS-49757], the
second fetch in the checkout step is avoided for cases where the
second fetch is not adding any new repository information to the
existing git repository.

Enabling this switch will allow the user to preserve the second
fetch and remedy any compatibility issues.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code. If a checkbox or line does not apply to this pull request, delete it. We prefer all checkboxes to be checked before a pull request is merged_

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

rishabhBudhouliya commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1483,6 +1482,7 @@ public ChangeLogParser createChangeLogParser() {
         private boolean useExistingAccountWithSameEmail;
 //        private GitClientType defaultClientType = GitClientType.GITCLI;
         private boolean showEntireCommitSummaryInChanges;
+        private boolean enablePerformanceImprovement = true; // By default, performance is enabled in git plugin
```
This branch was originally created to add performance improvement global switch. This switch has been used to add "opt out" feature for retaining second fetch during the checkout step.

rishabhBudhouliya commented on src/main/resources/hudson/plugins/git/GitSCM/global.jelly:
```diff
@@ -16,6 +16,9 @@
     <f:entry field="showEntireCommitSummaryInChanges">
       <f:checkbox title="${%Show the entire commit summary in changes}" name="showEntireCommitSummaryInChanges" checked="${descriptor.showEntireCommitSummaryInChanges}"/>
     </f:entry>
+    <f:entry field="enablePerformanceImprovement">
```
This has been changed to `allowSecondFetch` in the subsequent commits.

MarkEWaite commented on README.adoc:
```diff
@@ -166,6 +166,11 @@ Show the entire commit summary in changes::
+[[preserve-second-fetch-during-checkout]]
+Preserve second fetch during checkout::
+
+  If checked the checkout step will not avoid the second fetch.
```
```suggestion
  If checked the checkout step will not avoid the second fetch.
  Git plugin versions prior to git plugin 4.4 would perform two fetch operations for each repository checkout.
  Git plugin 4.4 removes the redundant fetch operation.
  Enabling this option will restore the redundant fetch operation.
  This setting is only needed if there is a bug in the redundant fetch removal logic.
  If you enable this setting, please report a git plugin issue that describes why you needed to enable it.
```

MarkEWaite commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -835,7 +840,6 @@ public GitClient createClient(TaskListener listener, EnvVars environment, Run<?,
 
         String gitExe = getGitExe(n, listener);
         Git git = Git.with(listener, environment).in(ws).using(gitExe);
-
```
We have a large number of "hidden" pull requests that risk being much more difficult to evaluate if we make changes that are pure white space changes.

Would you be willing to undo this whitespace change?

rishabhBudhouliya replied:
Apologies for adding it, I am not sure how this happened.

MarkEWaite commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1642,6 +1649,12 @@ public void setUseExistingAccountWithSameEmail(boolean useExistingAccountWithSam
 
+        public boolean isRedundantFetchAllowed() { return allowSecondFetch; }
+
+        public void setAllowSecondFetch(boolean allowSecondFetch) {
```
There are methods in Jenkins that use introspection to identify the getters and setters.  The asymmetry between the name of the getter and the setter in this case will probably confuse anything that uses introspection (like configuration as code and JobDSL)

Since the UI uses `allowSecondFetch`, I think you might consider using `allowSecondFetch` to replace `redundantFetchAllowed`.

```suggestion
        public void setRedundantFetchAllowed(boolean allowSecondFetch) {
```

rishabhBudhouliya commented:
@MarkEWaite I was also interactively testing this change and I am also facing this issue. The checkbox option is not being persisted across multiple build jobs.

I'll debug the instance for RCA.

rishabhBudhouliya commented:
@MarkEWaite I have done two things: 
- Interactively tested the checkbox, now the selection is persisting in my local instance.
- Removed `redundant` and replaced it with `second`.
